debugger/gdb-sgx-plugin/thread.py

```diff
@@ -0,0 +1,280 @@
+import gdb
+import math
+import tempfile
+from typing import List
+
+DEBUG = False
+
+INSTANCE_DEBUG_EXT = None
+
+class ThreadCreateBreakpoint(gdb.Breakpoint):
+    def __init__(self):
+        """
+        Set a breakpoing @ kernel/thread.c:_run_thread, this should be right before myst_setjmp(&thread->jmpbuf)
+        Choosing this location because this is where the newly created thread about to run
+        We can get the new thread pid/tid here
+        """
+        super(ThreadCreateBreakpoint, self).__init__('kernel/thread.c:_run_thread', internal=False)
```

Comment:
Can you test with a newer version of gdb? Newer versions of gdb could break *before* a functions prologue in which case the arguments have not yet been initialized

Reply:
> We also need to make sure that the hook functions are compiled with debug information, and aren't optimized out.
> 
> Do we expect the debugger extension to work in a RelWithDebInfo build of Mystikos?

Seems like currently Mystikos doesn't' differentiate between Debug or Release build, so hook functions are not optimized out for now.

Reply:
> Can you test with a newer version of gdb? Newer versions of gdb could break _before_ a functions prologue in which case the arguments have not yet been initialized

Could you elaborate? Would change to readying `%rdi` helps in this case?

Reply:
> Seems like currently Mystikos doesn't' differentiate between Debug or Release build, so hook functions are not optimized out for now.

@asvrada to safeguard the hooks from being optimized out we could wrap them with `#pragma GCC optimize "-O2"`. Though I'm not sure if the callsite can still be optimized out.

Reply:
> Can you test with a newer version of gdb? Newer versions of gdb could break _before_ a functions prologue in which case the arguments have not yet been initialized

I tested with gdb `GNU gdb (Ubuntu 9.2-0ubuntu1~20.04.1) 9.2` and it doesn't work just as you described...

Reply:
> > Can you test with a newer version of gdb? Newer versions of gdb could break _before_ a functions prologue in which case the arguments have not yet been initialized
> 
> Could you elaborate? Would change to reading `%rdi` helps in this case?

Yes reading from registers would be the safest. We do that elsewhere too:
https://github.com/openenclave/openenclave/blob/df8cb0d1807e80e5f9148dcad3841050a78ed6ad/debugger/gdb-extension/gdb_sgx_plugin.py#L283

Reply:
> > Can you test with a newer version of gdb? Newer versions of gdb could break _before_ a functions prologue in which case the arguments have not yet been initialized
> 
> I tested with gdb `GNU gdb (Ubuntu 9.2-0ubuntu1~20.04.1) 9.2` and it doesn't work just as you described...

I don't remember the exact version now; but we have run into that behavior before.